plugin/src/main/java/org/opensearch/ml/processor/MLInferenceIngestProcessor.java

```diff
@@ -0,0 +1,253 @@
+     * when no output mapping provided, default to output as
+     * "inference_results" field (the same format as predict API)
+     */
+    private void process_predictions(
+        IngestDocument ingestDocument,
+        BiConsumer<IngestDocument, Exception> handler,
+        List<Map<String, String>> process_input,
+        List<Map<String, String>> process_output,
+        int i,
+        int round
+    ) {
+        if (i >= round && i != 0) {
+            handler.accept(ingestDocument, null);
+        } else {
+            Map<String, String> modelParameters = new HashMap<>();
+            if (mlModelUtil.getModel_config() != null) {
+                modelParameters.putAll(mlModelUtil.getModel_config());
+            }
+            // when no input mapping is provided, default to read all fields from documents as model input
+            if (round == 0) {
+                Set<String> documentFields = ingestDocument.getSourceAndMetadata().keySet();
+                for (String field : documentFields) {
+                    getMappedModelInputFromDocuments(ingestDocument, modelParameters, field, field);
+                }
+
+            } else {
+                Map<String, String> inputMapping = process_input.get(i);
+                for (Map.Entry<String, String> entry : inputMapping.entrySet()) {
+                    String originalFieldName = entry.getKey();
+                    String ModelInputFieldName = entry.getValue();
+
+                    getMappedModelInputFromDocuments(ingestDocument, modelParameters, originalFieldName, ModelInputFieldName);
+                }
+            }
+
+            ActionRequest request = getRemoteModelInferenceResult(modelParameters, mlModelUtil.getModel_id());
+
+            client.execute(MLPredictionTaskAction.INSTANCE, request, new ActionListener<>() {
+
+                @Override
+                public void onResponse(MLTaskResponse mlTaskResponse) {
+                    ModelTensorOutput modelTensorOutput = (ModelTensorOutput) mlTaskResponse.getOutput();
+
+                    if (process_output == null || process_output.isEmpty()) {
+                        appendFieldValue(modelTensorOutput, null, DEFAULT_OUTPUT_FIELD_NAME, ingestDocument);
+                    } else {
+                        Map<String, String> outputMapping = process_output.get(i);
+                        for (Map.Entry<String, String> entry : outputMapping.entrySet()) {
+                            String originalModelOutputFieldName = entry.getKey(); // response
+                            String newModelOutputFieldName = entry.getValue(); // inference_result
+                            appendFieldValue(modelTensorOutput, originalModelOutputFieldName, newModelOutputFieldName, ingestDocument);
+                        }
+                    }
+                    process_predictions(ingestDocument, handler, process_input, process_output, i + 1, round);
```

Comment:
Looks like we're sending requests sequentially here and this could cause performance issue, can you try to optimize this?

Reply:
we are passing over the actionListener in each prediction, so it's not blocking call. But later we can enhance it with batch inference.